### PR TITLE
Fix #2211: Clone metadata

### LINKED ISSUE #2211: Clone the metadata of nodes
Hello all,

I would know if you agree to clone also the metadata when we clone a CtElement.

The use case is the following:

I find a specific element inside a method, _e.g._ a literal. According to its value, I generate multiple variants of the same method with different values for this literal.

To do this, I apply the following algorithm:
1. replace the value of the original literal with one of the generated
2. clone the CtMethod
3. restore the original value in the original method.

And I obtain a set of variants of the original method

This allows me to not execute the same query on the variants, _i.e_ the query to find specifc nodes.

However, I would like to add metadata on the literal, with the new values(_i.e._ before the step 2) but currently, it cannot be done using the `putMetadata()` method, since the metadata will be lost when cloning the method (at the step 2).

WDYT?

Best!

### PR DESCRIPTION
**Pull request by danglotb**

### CODE REVIEW

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 1]
```diff
@@ -469,6 +469,17 @@ public void replace(CtElement element) {
...
 	}
 
+	@Override
+	public <E extends CtElement> E setMetadata(Map<String, Object> metadata) {
+		if (metadata != null) {
+			if (this.metadata == null) {
+				this.metadata = new HashMap<>();
+			}
+			this.metadata.putAll(metadata);
+		}
```
Comment (pvojtechovsky) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: `else this.metadata = null`

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 2]
```diff
@@ -486,6 +497,11 @@ public Object getMetadata(String key) {
 		return metadata.get(key);
 	}
 
+	@Override
+	public Map<String, Object> getMetadata() {
+		return this.metadata;
```
Comment (pvojtechovsky) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: `return Collections.unmodifiableMap(this.metadata)`
WDYT?

**Comment by danglotb** [comment]
Thank you @pvojtechovsky for the review. 

I excluded `CtElement#setMetadata(Map<String, Object>)` from the test `AstCheckerTest#testPushToStackChanges()`.

WDYT?

File: src/test/java/spoon/generating/CloneVisitorGenerator.java [hunk 3]
```diff
@@ -212,7 +212,7 @@ private CtInvocation createFactoryInvocation(CtVariableAccess<CtElement> element
 					"spoon.support.reflect.declaration.CtModifiableImpl", "spoon.support.reflect.declaration.CtMultiTypedElementImpl", //
 					"spoon.support.reflect.declaration.CtTypeMemberImpl", "spoon.support.reflect.code.CtRHSReceiverImpl",
 					"spoon.support.reflect.declaration.CtShadowableImpl", "spoon.support.reflect.code.CtBodyHolderImpl", "spoon.support.reflect.declaration.CtModuleDirectiveImpl");
-			private final List<String> excludesFields = Arrays.asList("factory", "elementValues", "target", "metadata");
+			private final List<String> excludesFields = Arrays.asList("factory", "elementValues", "target");//, "metadata");
```
Comment (pvojtechovsky) [inline src/test/java/spoon/generating/CloneVisitorGenerator.java]: why that comment? It makes no sense for future. I would prefer to have the sources clean. WDYT?

**Comment by pvojtechovsky** [comment]
It looks good to me.
@monperrus, do you agree to add `CtElement#get/setMetadata(Map)` method?

**Comment by monperrus** [comment]
I propose to call them setAllMetadata/getAllMetadata

Otherwise, it's OK for me.

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 4]
```diff
@@ -469,6 +469,19 @@ public void replace(CtElement element) {
 		ReplacementVisitor.replace(this, elements);
 	}
 
+	@Override
+	public <E extends CtElement> E setMetadata(Map<String, Object> metadata) {
+		if (metadata != null) {
+			if (this.metadata == null) {
+				this.metadata = new HashMap<>();
+			}
+			this.metadata.putAll(metadata);
```
Comment (surli) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: You should have cleared the map first, else it's not a setter.

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 5]
```diff
@@ -469,6 +469,19 @@ public void replace(CtElement element) {
 		ReplacementVisitor.replace(this, elements);
 	}
 
+	@Override
+	public <E extends CtElement> E setMetadata(Map<String, Object> metadata) {
```
Comment (surli) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: question for @monperrus and @pvojtechovsky : shouldn't we call the model change listener here? Do we consider that metadata are not part of the model and the change on them shouldn't be notified?

Comment (pvojtechovsky) [reply src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: I would say change listener doesn't have to call change listener. But if one will needed. It is no problem to implement it later.

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 6]
```diff
@@ -486,6 +501,14 @@ public Object getMetadata(String key) {
 		return metadata.get(key);
 	}
 
+	@Override
+	public Map<String, Object> getAllMetadata() {
+		if (this.metadata == null) {
+			return null;
```
Comment (pvojtechovsky) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: what about `return Collections.emptyMap()`?

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 7]
```diff
@@ -469,6 +469,21 @@ public void replace(CtElement element) {
 		ReplacementVisitor.replace(this, elements);
 	}
 
+	@Override
+	public <E extends CtElement> E setAllMetadata(Map<String, Object> metadata) {
+		if (metadata != null) {
```
Comment (pvojtechovsky) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: `if (metadata != null && metadata.size() > 0) ...`
WDYT?

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java (hunk 5 above)
Comment (surli) [reply src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: ok then I just wanted to be sure it was not something missing.

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java (hunk 7 above)
Comment (surli) [reply src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: or this other way:
```
if (metadata == null) {
  // empty map
  return this;
}
this.metadata.clear();
this.metadata.putAll(metadata);
```

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 8]
```diff
@@ -469,6 +469,21 @@ public void replace(CtElement element) {
 		ReplacementVisitor.replace(this, elements);
 	}
 
+	@Override
+	public <E extends CtElement> E setAllMetadata(Map<String, Object> metadata) {
+		if (metadata == null || metadata.isEmpty()) {
+			this.metadata = new HashMap<>();
```
Comment (pvojtechovsky) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: empty hashmap has some size in memory and if we would have such empty hash maps more often (for each CtElement) then Spoon model would be n-times bigger - for nothing. 

I would prefer to use null to represent "no metadata attribute". WDYT?

File: src/main/java/spoon/reflect/declaration/CtElement.java [hunk 9]
```diff
@@ -305,6 +311,11 @@
 	 */
 	Object getMetadata(String key);
 
+	/**
+	 * Retrieves all metadata stored in an element. Returns null if it does not exist.
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/declaration/CtElement.java]: Now the comment needs update. It doesn't return null

**Comment by pvojtechovsky** [comment]
@surli, I am sorry, but I by mistake pressed ENTER during editing of commit message comment ... so I committed wrong message ... :-(

**Comment by surli** [comment]
No problem @pvojtechovsky ;)

**Pull request merged**